pkg/sethandler/controller.go

```diff
@@ -0,0 +1,227 @@
+			return setHandler.poolConfig.SelectPool(resNameAsString), nil
+		} else if strings.Contains(resNameAsString, resourceBaseName) && strings.Contains(resNameAsString, "exclusive") {
+			return setHandler.getListOfAllocatedExclusiveCpus(resNameAsString, pod, container)
+		}
+	}
+	return setHandler.poolConfig.SelectPool(resourceBaseName + "/default"), nil
+}
+
+func (setHandler *SetHandler) getListOfAllocatedExclusiveCpus(exclusivePoolName string, pod v1.Pod, container v1.Container) (types.Pool, error) {
+	checkpoint, err := checkpoint.GetCheckpoint()
+	if err != nil {
+		return types.Pool{}, errors.New("Kubelet checkpoint file could not be accessed because:" + err.Error())
+	}
+	podIDStr := string(pod.ObjectMeta.UID)
+	devices, err := checkpoint.GetComputeDeviceMap(podIDStr)
+	if err != nil {
+		return types.Pool{}, errors.New("List of assigned devices could not be read from checkpoint file for Pod: " + podIDStr + " because:" + err.Error())
+	}
+	exclusiveCpus, exist := devices[exclusivePoolName]
+	if !exist {
+		log.Println("WARNING: Container: " + container.Name + " in Pod: " + podIDStr + " asked for exclusive CPUs, but were not allocated any! Cannot adjust its default cpuset")
+		return types.Pool{}, nil
+	}
+	return calculateFinalExclusiveSet(exclusiveCpus, pod, container)
+}
+
+func calculateFinalExclusiveSet(exclusiveCpus *multusTypes.ResourceInfo, pod v1.Pod, container v1.Container) (types.Pool, error) {
+	var finalCPUSet types.Pool
+	var doesSetContainPinnerCore bool
+	for _, deviceID := range exclusiveCpus.DeviceIDs {
+		finalCPUSet.CPUs = finalCPUSet.CPUs + deviceID + ","
+		if deviceID == dedicatedPinnerCoreID {
+			doesSetContainPinnerCore = true
+		}
+	}
+	if isPinnerUsedByContainer(pod, container.Name) && !doesSetContainPinnerCore {
+		//1: Container is asking exclusive CPUs + 2: Container has a pool configured in the annotation = Pinner will be used
+		//If pinner is used by a container, we need to the add the configured CPU core holding its thread to their cpuset
+		finalCPUSet.CPUs = finalCPUSet.CPUs + dedicatedPinnerCoreID + ","
+	}
+	finalCPUSet.CPUs = strings.TrimSuffix(finalCPUSet.CPUs, ",")
+	return finalCPUSet, nil
+}
+
+func isPinnerUsedByContainer(pod v1.Pod, containerName string) bool {
+	for key, value := range pod.ObjectMeta.Annotations {
+		if strings.Contains(key, processConfigKey) {
+			var processConfig types.CPUAnnotation
+			err := json.Unmarshal([]byte(value), &processConfig)
+			if err != nil {
+				return false
+			}
+			containerConfig := processConfig.ContainerPools(containerName)
+			if len(containerConfig) > 0 {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func determineCid(podStatus v1.PodStatus, containerName string) string {
+	for _, containerStatus := range podStatus.ContainerStatuses {
+		if containerStatus.Name == containerName {
+			return containerStatus.ContainerID
+		}
+	}
+	return ""
+}
+
+func (setHandler *SetHandler) applyCpusetToContainer(containerID string, cpuset types.Pool) error {
```

Comment:
cpuset and and type for it do not correlate well. The cpuset type could be just a string as the parameter is not handling a pool ? There are other similar cases in this file too.

Reply:
well, the Pool type effectively has only one member, called CPUs, so the existing type in reality is a cpuset :)

I wanted to use an abstract type throughout the methods and interfaces, to make it extensible in the future if we figure out we need to store something else in it, or we want to implement some functions on it.

As the existing type looked exactly what I wanted I thought to re-use, because then I can just simply pass these Pools to the underlying layer from SetHandler (e.g. return setHandler.poolConfig.SelectPool(resourceBaseName + "/default")), without needing to maintain another, intermediate representation of the data.

But if you want I can introduce my own CPUSet type, and usethat throughout the code.

Reply:
https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/pkg/kubelet/cm/cpuset/cpuset.go
could re-use this type for example

Reply:
That would be good if it suits for the purpose and is not too difficult to use.

Reply:
I was actually thinking if it would be better to remove the Pool type and put the CPUs directly to the map (PoolConfig.pools).

Reply:
yeah, my original dataset was actually three []int slices
ok, anyway, I changed the Pool to cpuset.CPUSet type  in the insides of the controller
SetHandler still holds PoolConfigs as member, but they are transformed into CPUSet types as soon as possible
It is actually quite nice type, easy to work with